### PR TITLE
Added support for final, enum abstract and arrow functions

### PR DESCRIPTION
**Pull request by AlexHaxe**
Not sure if everything works, but at least the tests pass.

### CODE REVIEW

**Comment by Simn** [comment]
Nice! Let me know when you're done.

**Comment by AlexHaxe** [comment]
I recon with these patches haxeparser is more or less up to date. There could still be some changes I might have overlooked - I haven't done an in-depth comparison of grammar.mly <-> HaxeParser.hx. But these can be handled in additional PRs.

I skipped all that Not-XML stuff, since it's not really final.

I think I am done for now, unless you find something that needs a better implementation.

File: src/haxeparser/HaxeParser.hx [hunk 1]
```diff
@@ -826,10 +949,17 @@ class HaxeParser extends hxparse.Parser<HaxeTokenSource, Token> implements hxpar
 			case [{tok:Arrow}, t2 = parseComplexType()]:
 				switch(t2) {
 					case TFunction(args,r):
-						TFunction(apush(args,t),r);
+						TFunction(aunshift(args,t),r);
```
Comment (Simn) [inline src/haxeparser/HaxeParser.hx]: Hmm, what's the story here?

Comment (AlexHaxe) [reply src/haxeparser/HaxeParser.hx]: `var f : (Int, Float) -> Int;` would print as `var f : (Float, Int) -> Int;` in unittest, so I looked it up and it's `t :: args` which means it should be unshift not push (as far as my understanding of ocaml goes)

**Pull request merged**